plugin/plugin.lua

```diff
@@ -0,0 +1,268 @@
+--
+-- The main plugin implementation that can run in an NGINX system with the LUA module enabled
+--
+
+local _M = { conf = {} }
+
+local base64 = require 'ngx.base64'
+local cipher = require 'resty.openssl.cipher'
+
+local VERSION_SIZE    = 1
+local GCM_IV_SIZE     = 12
+local GCM_TAG_SIZE    = 16
+local CURRENT_VERSION = 1
+
+local function array_has_value(arr, val)
+    for index, value in ipairs(arr) do
+        if value == val then
+            return true
+        end
+    end
+    return false
+end
+
+local function from_hex(str)
+    return (str:gsub('..', function (cc)
+        return string.char(tonumber(cc, 16))
+    end))
+end
+
+local function get_csrf_header_name(config)
+    return 'x-' .. config.cookie_name_prefix .. '-csrf'
+end
+
+local function apply_configuration_defaults(config)
+
+    if not config.allow_tokens then
+        config.allow_tokens = false
+    end
+
+    if not config.remove_cookie_headers then
+        config.remove_cookie_headers = true
+    end
+
+    if config.cors_enabled then
+
+        if not config.cors_allowed_methods then
+            config.cors_allowed_methods = { 'OPTIONS', 'GET', 'HEAD', 'POST', 'PUT', 'PATCH', 'DELETE'}
+        end
+
+        if not config.cors_allowed_headers then
+            config.cors_allowed_headers = { get_csrf_header_name(config) }
+        end
+
+        if not config.cors_exposed_headers then
+            config.cors_exposed_headers = {}
+        end
+
+        if not config.cors_max_age then
+            config.cors_max_age = 86400
+        end
+    end
+end
+
+local function add_cors_response_headers(config)
+
+    if config.trusted_web_origins then
```

Comment:
Changed this so that the apply config defaults code sets it to an empty array when not supplied. This means the rest of the code does not need to check for null.